### PR TITLE
Add scripts to configure Elasticsearch data source

### PR DESCRIPTION
**Pull request by cumcke**
- Created two scripts to configure the Elasticsearch data source in Grafana (one for cluster and one for tenant).
- Currently, the tenant admin that's created with the logging onboarding scripts do not have enough permission so, for the time being, we either need to either:
- - Update the permission of the tenant admin user to include **indices:admin/mappings/get** for **viya_logs-* ** in order for it to work as intended (will be creating new users for these scripts in a later release.
- - Use admin user ID.

I also used this opportunity to update a multi-tenant Prometheus YAML file to include the correct URL (v4m-prometheus-__TENANT__:9090).

### CODE REVIEW

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+source monitoring/bin/common.sh
+export LOG_NS="${LOG_NS:-logging}"
+
+# Check to see if monitoring namespace provided exists and components have already been deployed
+log_info "Checking for Grafana pods in the $MON_NS namespace ..."
+if [[ $(kubectl get pods -n $MON_NS -l app.kubernetes.io/instance=v4m-prometheus-operator -o custom-columns=:metadata.namespace --no-headers | uniq | wc -l) == 0 ]]; then
+  log_error "No monitoring components found in the $MON_NS namespace.  Monitoring needs to be deployed in this namespace in order to configure Elasticsearch with this script.";
+  exit 1
+else
+  log_verbose "Monitoring found in $MON_NS namespace.  Continuing."
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:22]: Seems like this should be a log_debug message

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+fi
+
+# Check to see if logging namespace provided exists and components have already been deployed
+log_info "Checking for Elasticsearch pods in the $LOG_NS namespace ..."
+if [[ $(kubectl get pods -A -l app=v4m-es -o custom-columns=:metadata.namespace --no-headers | uniq | wc -l) == 0 ]]; then
+  log_error "No logging components found in the $LOG_NS namespace.  Logging needs to be deployed in this namespace in order to configure Elasticsearch with this script.";
+else
+  log_verbose "Logging found in $LOG_NS namespace.  Continuing."
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:30]: As above, should this be log_debug?

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+fi
+
+# Create temporary directory for string replacement in the grafana-datasource-es.yaml file
+  monDir=$TMP_DIR/$MON_NS
+  mkdir -p $monDir
+  cp -R monitoring/grafana-datasource-es.yaml $monDir/grafana-datasource-es.yaml
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:36]: don't need -R

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+# Pull Elasticsearch password in temp directory file
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:38]: "in temp directory file" is incorrect here. It's just an in-memory variable

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+adminPass="$(kubectl -n $LOG_NS get secret internal-user-admin -o=jsonpath="{.data.password}" | base64 --decode)"
+
+# Replace placeholders
+log_debug "Replacing logging namespace for files in [$monDir]"
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:42]: Suggested message: "Replacing variables in $monDir/grafana-datasource-es.yaml"

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+fi
+  
+log_info "Provisioning Elasticsearch datasource for Grafana"
+kubectl delete secret -n $MON_NS --ignore-not-found grafana-datasource-es
+kubectl create secret generic -n $MON_NS grafana-datasource-es --from-file $monDir/grafana-datasource-es.yaml
+kubectl label secret -n $MON_NS grafana-datasource-es grafana_datasource=1 sas.com/monitoring-base=kube-viya-monitoring
+
+# Delete pods so that they can be restarted with the change.
+iog_info "Elasticsearch data source provisioned.  Restarting pods to apply the change"
+kubectl delete pods -n $tenantNS -l "app.kubernetes.io/instance=v4m-grafana-acme"
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:62]: This is the cluster script, so it should be `-n $MON_NS -l "app.kubernetes.io/instance=v4m-grafana"`

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+kubectl -n $tenantNS wait pods --selector app.kubernetes.io/instance=v4m-grafana-acme --for condition=Ready --timeout=2m
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:63]: Same thing - use `$MON_NS` and `v4m-grafana`

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+done
+
+# Convert namespace and tenant to all lower-case
+tenantNS=$(echo "$tenantNS"| tr '[:upper:]' '[:lower:]')
+tenant=$(echo "$tenant"| tr '[:upper:]' '[:lower:]')
+
+# Check to see if monitoring deployment could be found for tenant in provided namespace
+log_info "Checking the $tenantNS namespace for monitoring deployment for the $tenant tenant ..."
+if [[ $(kubectl get pods -n $tenantNS -l app.kubernetes.io/instance=v4m-grafana-$tenant -o custom-columns=:metadata.namespace --no-headers | uniq | wc -l) == 0 ]]; then
+    log_error "No monitoring components found in the $tenantNS namespace.  Monitoring needs to be deployed in this namespace in order to configure Elasticsearch with this script.";
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:100]: Split into two log_error messages on the sentence boundary.

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    exit 1
+else
+    log_verbose "Monitoring deployment was found for the $tenant tenant in the $tenantNS namespace.  Continuing."
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:103]: log_debug

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+fi
+
+# Verify that the password is correct?
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:106]: Either implement the check or remove the comment

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+
+# Create temporary directory for string replacement in the grafana-datasource-es.yaml file
+tenantDir=$TMP_DIR/$VIYA_TENANT
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:110]: There is no `$VIYA_TENANT` variable here - use `$tenant`

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+mkdir -p $tenantDir
+cp -R monitoring/grafana-datasource-es.yaml $tenantDir/grafana-datasource-es.yaml
+
+# Replace placeholders
+log_debug "Replacing logging namespace for files in [$tenantDir]"
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:115]: Change message to `Replacing variables in $tenantDir/grafana-datasource-es.yaml`

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  sed -i '' "s/__namespace__/$LOG_NS/g" $tenantDir/grafana-datasource-es.yaml
+  sed -i '' "s/__userID__/$userID/g" $tenantDir/grafana-datasource-es.yaml
+  sed -i '' "s/__passwd__/$passwd/g" $tenantDir/grafana-datasource-es.yaml
+else
+  sed -i "s/__namespace__/$LOG_NS/g" $tenantDir/grafana-datasource-es.yaml
+  sed -i "s/__userID__/$userID/g" $tenantDir/grafana-datasource-es.yaml
+  sed -i "s/__passwd__/$passwd/g" $tenantDir/grafana-datasource-es.yaml
+fi
+
+log_info "Provisioning Elasticsearch data source for Grafana"
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:126]: Message: Provisioning Elasticsearch data source for [$tenantNS/$tenant] instance of Grafana

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+kubectl delete secret -n $tenantNS --ignore-not-found v4m-grafana-datasource-es-$tenant
+kubectl create secret generic -n $tenantNS v4m-grafana-datasource-es-$tenant --from-file $tenantDir/grafana-datasource-es.yaml
+kubectl label secret -n $tenantNS v4m-grafana-datasource-es-$tenant grafana_datasource-$tenant=true sas.com/monitoring-base=kube-viya-monitoring
+
+# Delete pods so that they can be restarted with the change.
+iog_info "Elasticsearch data source provisioned.  Restarting pods to apply the change"
+kubectl delete pods -n $tenantNS -l "app.kubernetes.io/instance=v4m-grafana-acme"
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:133]: Hard-coded `acme` here. Use `$tenant`

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+kubectl -n $tenantNS wait pods --selector app.kubernetes.io/instance=v4m-grafana-acme --for condition=Ready --timeout=2m
```
Comment (BryanEllington) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:134]: Replace hard-coded `acme`

**Review by BryanEllington — CHANGES_REQUESTED** [review]

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    log_verbose "Monitoring deployment was found for the $tenant tenant in the $tenantNS namespace.  Continuing."
+fi
+
+# Verify that the password is correct?
+
+
+# Create temporary directory for string replacement in the grafana-datasource-es.yaml file
+tenantDir=$TMP_DIR/$VIYA_TENANT
+mkdir -p $tenantDir
+cp -R monitoring/grafana-datasource-es.yaml $tenantDir/grafana-datasource-es.yaml
```
Comment (gsmith-sas) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:112]: Another unneeded -R on a cp command.

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+  log_error "No monitoring components were found for $tenant in the $tenantNS namespace."
+  log_error "Monitoring needs to be deployed in this namespace in order to configure Elasticsearch with this script.";
+  exit 1
+else
+  log_debug "Monitoring deployment was found for the $tenant tenant in the $tenantNS namespace.  Continuing."
+fi
+
+# Check to see if tenant has been deployed in logging.
+log_info "Checking that $tenant has been onboarded to Elasticsearch in the $LOG_NS namespace ..."
+if [[ $(kibana_tenant_exists $tenant) == 1 ]]; then
```
Comment (gsmith-sas) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:130]: You can simplify this syntax to:
`if ! kibana_tenant_exists "$ktenant"; then`

More importantly, I don't see a call to the get_sec_api function; this function must be called to establish the connection to Elasticsearch.  Is it being called somewhere else?

File: monitoring/bin/create_elasticsearch_datasource_cluster.sh [hunk 18]
```diff
@@ -16,50 +16,55 @@ export LOG_NS="${LOG_NS:-logging}"
...
 else
   sed -i "s/__sourceName__/Elastic/g" $monDir/grafana-datasource-es.yaml
-  sed -i "s/__namespace__/$logNS/g" $monDir/grafana-datasource-es.yaml
+  sed -i "s/__namespace__/$LOG_NS/g" $monDir/grafana-datasource-es.yaml
   sed -i "s/__userID__/admin/g" $monDir/grafana-datasource-es.yaml
   sed -i "s/__passwd__/$adminPass/g" $monDir/grafana-datasource-es.yaml
 fi
-  
+
+if [ ! -z "$(kubectl get secret -n $MON_NS grafana-datasource-es -o custom-columns=:metadata.name --no-headers --ignore-not-found)"]; then
```
Comment (gsmith-sas) [inline monitoring/bin/create_elasticsearch_datasource_cluster.sh:57]: Mostly a "style" thing but you could replace `! -z` with `-n`...but we haven't adopted formal coding conventions or standardization.

**Review by gsmith-sas — CHANGES_REQUESTED** [review]
See my comment about the get_sec_api function...it doesn't seem to be called and I don't know how the tenant exist function call can work without it.

File: monitoring/bin/create_elasticsearch_datasource_tenant.sh [hunk 19]
```diff
@@ -127,7 +129,11 @@ fi
 
 # Check to see if tenant has been deployed in logging.
 log_info "Checking that $tenant has been onboarded to Elasticsearch in the $LOG_NS namespace ..."
-if [[ $(kibana_tenant_exists $tenant) == 1 ]]; then
+
+get_sec_api_url
+get_credentials_from_secret admin
+
+if ! kibana_tenant_exists "${tenantNS}_${tenant}"; then
   log_error "Unable to configure Elasticsearch data source.  The $tenant tenant has not been onboarded.";
```
Comment (gsmith-sas) [inline monitoring/bin/create_elasticsearch_datasource_tenant.sh:137]: Probably should include the "namespace" in the error message as well.  e.g.  _The tenant [foo] within the namespace [bar] has not been onboarded._

**Review by gsmith-sas — CHANGES_REQUESTED** [review]
A small tweak to an error message but otherwise looks good.

**Pull request merged**